daemons/ipa-slapi-plugins/ipa-extdom-extop/back_extdom_nss_sss.c

```diff
@@ -0,0 +1,278 @@
+	ctx->initgroups_dyn = dlsym(ctx->dl_handle, "_nss_sss_initgroups_dyn");
+	if (ctx->initgroups_dyn == NULL) {
+		goto fail;
+	}
+
+	return 0;
+
+fail:
+	back_extdom_free_context(nss_context);
+
+	return -1;
+}
+
+
+/* Following three functions cannot be implemented with nss_sss.so.2
+ * As result, we simply do nothing here */
+
+void back_extdom_set_timeout(struct nss_ops_ctx *nss_context,
+			     unsigned int timeout) {
+	/* no operation */
+}
+
+void back_extdom_evict_user(struct nss_ops_ctx *nss_context,
+			    const char *name) {
+	/* no operation */
+}
+
+void back_extdom_evict_group(struct nss_ops_ctx *nss_context,
+			     const char *name) {
+	/* no operation */
+}
+
+#include <stdio.h>
+enum nss_status back_extdom_getpwnam(struct nss_ops_ctx *nss_context,
+				     const char *name, struct passwd *pwd,
+				     char *buffer, size_t buflen,
+				     struct passwd **result,
+				     int *lerrno) {
+	enum nss_status ret;
+
+	if (nss_context == NULL) {
+		return NSS_STATUS_UNAVAIL;
+	}
+
+	ret = (enum nss_status)
+		nss_context->getpwnam_r(name, pwd,
+					buffer, buflen,
+					lerrno);
+
+	if ((ret == NSS_STATUS_SUCCESS) && (result != NULL)) {
+		*result = pwd;
+		*lerrno = 0;
+	}
+
+	return ret;
+}
+
+enum nss_status back_extdom_getpwuid(struct nss_ops_ctx *nss_context,
+				     uid_t uid, struct passwd *pwd,
+				     char *buffer, size_t buflen,
+				     struct passwd **result,
+				     int *lerrno) {
+	enum nss_status ret;
+
+	if (nss_context == NULL) {
+		return NSS_STATUS_UNAVAIL;
+	}
+
+	ret = (enum nss_status)
+		nss_context->getpwuid_r(uid, pwd,
+					buffer, buflen,
+					lerrno);
+
+	if ((ret == NSS_STATUS_SUCCESS) && (result != NULL)) {
+		*result = pwd;
+	}
+
+	return ret;
+}
+
+enum nss_status back_extdom_getgrnam(struct nss_ops_ctx *nss_context,
+				     const char *name, struct group *grp,
+				     char *buffer, size_t buflen,
+				     struct group **result,
+				     int *lerrno) {
+	enum nss_status ret;
+
+	if (nss_context == NULL) {
+		return NSS_STATUS_UNAVAIL;
+	}
+
+	ret = (enum nss_status)
+		nss_context->getgrnam_r(name, grp,
+					buffer, buflen,
+					lerrno);
+
+	if ((ret == NSS_STATUS_SUCCESS) && (result != NULL)) {
+		*result = grp;
+	}
+
+	return ret;
+}
+
+enum nss_status back_extdom_getgrgid(struct nss_ops_ctx *nss_context,
+				     gid_t gid, struct group *grp,
+				     char *buffer, size_t buflen,
+				     struct group **result,
+				     int *lerrno) {
+
+	enum nss_status ret;
+
+	if (nss_context == NULL) {
+		return NSS_STATUS_UNAVAIL;
+	}
+
+	ret = (enum nss_status)
+		nss_context->getgrgid_r(gid, grp,
+					buffer, buflen,
+					lerrno);
+
+	if ((ret == NSS_STATUS_SUCCESS) && (result != NULL)) {
+		*result = grp;
+	}
+
+	return ret;
+}
+
+enum nss_status back_extdom_getgrouplist(struct nss_ops_ctx *nss_context,
+					 const char *name, gid_t group,
+					 gid_t *groups, int *ngroups,
+					 int *lerrno) {
+
+	enum nss_status ret = NSS_STATUS_UNAVAIL;
+	long int tsize = MAX (1, *ngroups);
```

Comment:
Should we have a ceiling on how much memory we are willing to allocate ?
Do we need to check for negative numbers here ?
can we define ngroups as unsigned ?

Reply:
Not handled

Reply:
ngroups is int in the nss interface -- see `getgrouplist(3)`. It is one of few areas where glibc and POSIX aren't consistent -- `initgroups_dyn()` API requires `long int` for number of groups while `getgrouplist()` which is an upper level API for `initgroups_dyn()` uses `int`.

Reply:
We don't need to check on negative numbers because we do have `MAX(1, *ngroups)`, so we are guaranteed at least `1` here.

Also, the calls to `back_extdom_getgrouplist()` have no influence from external sources. Namely, `get_user_grouplist()` in ipa_extdom_common.c sets ngroups to 128 and does not depend on any user input.